shared-types.d.ts

```diff
@@ -0,0 +1,274 @@
+ */
+export interface Effects {
+  /**
+   * Enter and exit define where tokens start and end
+   */
+  enter: (type: Type) => void
+
+  /**
+   * Enter and exit define where tokens start and end
+   */
+  exit: (type: Type) => void
+
+  /**
+   * Consume deals with a character, and moves to the next
+   */
+  consume: (code: number) => void
+
+  /**
+   * Attempt deals with several values, and tries to parse according to those values.
+   * If a value resulted in `ok`, it worked, the tokens that were made are used,
+   * and `returnState` is switched to.
+   * If the result is `nok`, the attempt failed,
+   * so we revert to the original state, and `bogusState` is used.
+   */
+  attempt: (
+    constructInfo:
+      | Construct
+      | Construct[]
+      | {[code: number]: Construct | Construct[]},
+    returnState: State,
+    bogusState?: State
+  ) => (code: number) => void
+
+  /**
+   * interrupt is used for stuff right after a line of content.
+   */
+  interrupt: (
+    constructInfo:
+      | Construct
+      | Construct[]
+      | {[code: number]: Construct | Construct[]},
+    ok: Okay,
+    nok?: NotOkay
+  ) => (code: number) => void
+
+  check: (
+    constructInfo:
+      | Construct
+      | Construct[]
+      | {[code: number]: Construct | Construct[]},
+    ok: Okay,
+    nok?: NotOkay
+  ) => (code: number) => void
+
+  /**
+   * lazy is used for lines that were not properly preceded by the container.
+   */
+  lazy: (
+    constructInfo:
+      | Construct
+      | Construct[]
+      | {[code: number]: Construct | Construct[]},
+    ok: Okay,
+    nok?: NotOkay
+  ) => void
+}
+
+/**
+ * A state function should return another function: the next state-as-a-function to go to.
+ *
+ * But there is one case where they return void: for the eof character code (at the end of a value)
+ * The reason being: well, there isn’t any state that makes sense, so void works well. Practically
+ * that has also helped: if for some reason it was a mistake, then an exception is throw because
+ * there is no next function, meaning it surfaces early.
+ */
+export type State = (code: number) => State | void
+
+/**
+ *
+ */
+export type Okay = State
+
+/**
+ *
+ */
+export type NotOkay = State
+
+/**
+ *
+ */
+export interface Tokenizer {
+  previous: Code
+  events: Event[]
+  parser: Parser
+  sliceStream: (token: Token) => Chunk[]
+  sliceSerialize: (token: Token) => string
+  now: () => Point
+  defineSkip: (value: Point) => void
+  write: (slice: Chunk[]) => Event[]
+}
+
+export type Resolve = (events: Event[], context: Tokenizer) => Event[]
+
+export type Tokenize = (context: Tokenizer, effects: Effects) => State
+
+export interface Construct {
+  tokenize: Tokenize
+  partial?: boolean
+  resolve?: Resolve
+  resolveTo?: Resolve
+  resolveAll?: Resolve
+  concrete?: boolean
+  interruptible?: boolean
+  lazy?: boolean
+}
+
+/**
+ *
+ */
+export interface Parser {
+  constructs: Record<NonNullable<Code>, Construct | Construct[]>
```

Comment:
To continue the conversation: https://github.com/micromark/micromark/pull/25#discussion_r499209678

It is indeed a `String(code)`, which is done implicitly when setting `null` or `number` on an object by JavaScript.
The key EOF code `null` is thus `"null"`

Reply:
changed the type here to `NonNullable<Code> | 'null'`